File: python/tests/test_golden.py
"""Golden tests — same cases as tests/golden/cases (C++ suite)."""

from __future__ import annotations

from pathlib import Path

import pytest

from mcrpc import (
    DiscoverBuilder,
    Dispatcher,
    StatusBuilder,
    build_call_err,
    build_call_ok,
    build_event,
    is_valid_proc,
    parse,
)
from mcrpc.request import ParseResult
from mcrpc.version import PROTOCOL_VERSION

GOLDEN_DIR = Path(__file__).resolve().parents[2] / "tests" / "golden" / "cases"


def _trim(s: str) -> str:
    return s.strip()


def load_cases(path: Path) -> list[dict]:
    cases: list[dict] = []
    cur: dict = {}
    st = None

    def flush() -> None:
        nonlocal cur
        if cur.get("in") is not None or cur.get("out") is not None:
            cur.setdefault("mode", "dispatch")
            cases.append(cur)
        cur = {}

    for line in path.read_text(encoding="utf-8").splitlines():
        if line.startswith("#") and (len(line) == 1 or line[1] in " \t"):
            continue
        t = _trim(line)
        if t == "---":
            flush()
            st = None
            continue
        if t == "IN":
            st = "in"
            continue
        if t == "OUT":
            st = "out"
            continue
        if t == "MODE":
            st = "mode"
            continue
        if st == "in":
            cur["in"] = t
            st = None
        elif st == "out":
            cur["out"] = t
            st = None
        elif st == "mode":
            cur["mode"] = t
            st = None
    flush()
    return cases


def _handle_call(req):
    if not req.args:
        return build_call_err("invalid_argument")
    proc = req.args[0]
    if not is_valid_proc(proc):
        return build_call_err("invalid_argument", reason="proc")
    if proc == "scene.nope":
        return build_call_err("unknown_proc")
    if proc == "button.pressed":
        return build_call_ok(lat="50", lon="19")
    return build_call_ok()


def _make_dispatcher() -> Dispatcher:
    d = Dispatcher()
    d.set_node_name("tracker")
    d.set_group_name("mych")
    d.register("ping", lambda req: "pong")
    d.register("call", _handle_call)
    d.register("relay", lambda req: "err unsupported")
    d.register("battery", lambda req: "err unsupported")
    d.register("gps", lambda req: "err unsupported")
    return d


@pytest.mark.parametrize("case_file", sorted(GOLDEN_DIR.glob("*.txt")), ids=lambda p: p.name)
def test_golden_file(case_file: Path) -> None:
    assert GOLDEN_DIR.is_dir(), f"golden dir missing: {GOLDEN_DIR}"
    disp = _make_dispatcher()
    for case in load_cases(case_file):
        mode = case.get("mode", "dispatch")
        if mode in ("dispatch", "ignore"):
            got = disp.dispatch(case["in"])
            if mode == "ignore":
                assert got is None, f"expected no reply for {case}"
                continue
            assert got == case["out"], f"{case_file.name}: IN={case['in']!r} EXP={case['out']!r} GOT={got!r}"
            continue
        if mode == "build_event":
            name = case["in"]
            kv = ""
            if " " in name:
                name, kv = name.split(" ", 1)
            got = build_event(name, kv or None)
            assert got == case["out"]
            continue
        if mode == "build_request":
            result, req = parse(case["in"])
            assert result == ParseResult.Ok
            rid = req.request_id if req.has_request_id else None
            from mcrpc import build_request

            got = build_request(req.target, req.command, request_id=rid)
            assert got == case["out"]
            continue
        if mode == "build_discover":
            b = DiscoverBuilder()
            b.set_node_name("tracker")
            b.add("id", "3cbbf74e")
            b.add("fw", "test")
            b.add("v", PROTOCOL_VERSION)
            b.add("tag", "tracker")
            b.add("up", "42s")
            b.add("caps", "battery,button")
            got = b.write()
            assert got == case["out"]
            continue
        if mode == "build_status":
            s = StatusBuilder()
            s.add("name", "tracker")
            s.add("id", "3cbbf74e")
            s.add("fw", "test")
            s.add("v", PROTOCOL_VERSION)
            s.add("tag", "tracker")
            s.add("transport", "meshcore")
            got = s.write()
            assert got.startswith(case["out"]) or got == case["out"]
            continue
        pytest.fail(f"unknown MODE {mode}")

File: tests/golden/test_golden.cpp
#include <mcrpc/mcrpc.h>
#include <cstdio>
#include <cstring>
#include <cstdlib>
#include <string>
#include <vector>
#include <fstream>
#include <sstream>
#include <filesystem>

using namespace mcrpc;
namespace fs = std::filesystem;

#ifndef MCRPC_GOLDEN_DIR
#define MCRPC_GOLDEN_DIR "tests/golden/cases"
#endif

static int g_fail = 0;
static int g_pass = 0;

static bool hPing(CommandContext& ctx) {
  ctx.reply->clear();
  ctx.reply->append("pong");
  return true;
}

static bool hUnsupported(CommandContext& ctx) {
  ctx.reply->clear();
  ctx.reply->append("err unsupported");
  return true;
}

struct Case {
  std::string in;
  std::string out;
  std::string mode;  // dispatch | ignore | build_event | build_request | build_discover | build_status
};

static std::string trim(std::string s) {
  while (!s.empty() && (s.back() == '\r' || s.back() == '\n' || s.back() == ' ')) s.pop_back();
  size_t i = 0;
  while (i < s.size() && (s[i] == ' ' || s[i] == '\t')) i++;
  return s.substr(i);
}

static std::vector<Case> loadFile(const fs::path& path) {
  std::ifstream in(path);
  std::vector<Case> cases;
  Case cur;
  std::string line;
  enum { None, WantIn, WantOut, WantMode } st = None;
  auto flush = [&]() {
    if (!cur.in.empty() || !cur.out.empty()) {
      if (cur.mode.empty()) cur.mode = "dispatch";
      cases.push_back(cur);
    }
    cur = Case{};
    st = None;
  };
  while (std::getline(in, line)) {
    // Comments: "# …" (hash + whitespace). Request-id bodies like "#18 pong" are data.
    if (!line.empty() && line[0] == '#' &&
        (line.size() == 1 || line[1] == ' ' || line[1] == '\t'))
      continue;
    std::string t = trim(line);
    if (t == "---") {
      flush();
      continue;
    }
    if (t == "IN") {
      st = WantIn;
      continue;
    }
    if (t == "OUT") {
      st = WantOut;
      continue;
    }
    if (t == "MODE") {
      st = WantMode;
      continue;
    }
    if (st == WantIn) {
      cur.in = t;
      st = None;
    } else if (st == WantOut) {
      cur.out = t;
      st = None;
    } else if (st == WantMode) {
      cur.mode = t;
      st = None;
    }
  }
  flush();
  return cases;
}

static void check(const char* file, const Case& c, const std::string& got, bool expect_reply) {
  if (!expect_reply) {
    if (c.out == "-" || c.out.empty()) {
      g_pass++;
      return;
    }
  }
  if (got == c.out) {
    g_pass++;
    return;
  }
  // soft prefix match for status (uptime/rssi may vary)
  if (c.mode == "build_status" && got.rfind(c.out, 0) == 0) {
    g_pass++;
    return;
  }
  std::printf("FAIL %s\n  IN:  %s\n  EXP: %s\n  GOT: %s\n", file, c.in.c_str(), c.out.c_str(),
              got.c_str());
  g_fail++;
}

static void runCase(const char* file, const Case& c, Dispatcher& disp) {
  ReplyBuffer buf;
  if (c.mode == "dispatch" || c.mode == "ignore") {
    bool replied = disp.dispatch(c.in.c_str(), buf);
    if (c.mode == "ignore") {
      if (!replied) {
        g_pass++;
        return;
      }
      std::printf("FAIL %s expected no reply, got '%s'\n", file, buf.data);
      g_fail++;
      return;
    }
    check(file, c, replied ? std::string(buf.data) : std::string(), true);
    return;
  }
  if (c.mode == "build_event") {
    // IN is "name [kv]"
    std::string name = c.in;
    std::string kv;
    auto sp = name.find(' ');
    if (sp != std::string::npos) {
      kv = name.substr(sp + 1);
      name = name.substr(0, sp);
    }
    OutboundBuilder::event(buf, name.c_str(), kv.empty() ? nullptr : kv.c_str());
    check(file, c, buf.data, true);
    return;
  }
  if (c.mode == "build_request") {
    // IN like tracker#42 status or tracker gps
    Request r;
    if (Parser::parse(c.in.c_str(), r) != ParseResult::Ok) {
      std::printf("FAIL %s parse for build_request\n", file);
      g_fail++;
      return;
    }
    OutboundBuilder::request(buf, r.target, r.command, r.has_request_id, r.request_id);
    check(file, c, buf.data, true);
    return;
  }
  if (c.mode == "build_discover") {
    DiscoverBuilder d;
    d.setNodeName("tracker");
    d.add("id", "3cbbf74e");
    d.add("fw", "test");
    d.add("v", protocolVersionString());
    d.add("tag", "tracker");
    d.add("up", "42s");
    d.add("caps", "battery,button");
    d.writeTo(buf);
    check(file, c, buf.data, true);
    return;
  }
  if (c.mode == "build_status") {
    StatusBuilder s;
    s.add("name", "tracker");
    s.add("id", "3cbbf74e");
    s.add("fw", "test");
    s.add("v", protocolVersionString());
    s.add("tag", "tracker");
    s.add("transport", "meshcore");
    s.writeTo(buf);
    check(file, c, buf.data, true);
    return;
  }
  std::printf("FAIL %s unknown MODE %s\n", file, c.mode.c_str());
  g_fail++;
}

static bool hCall(CommandContext& ctx) {
  if (ctx.request->argc < 1) {
    CallResult::err(*ctx.reply, "invalid_argument");
    return true;
  }
  const char* proc = ctx.request->args[0];
  if (!CallResult::isValidProc(proc)) {
    CallResult::err(*ctx.reply, "invalid_argument");
    CallResult::appendKv(*ctx.reply, "reason", "proc");
    return true;
  }
  if (std::strcmp(proc, "scene.nope") == 0) {
    CallResult::err(*ctx.reply, "unknown_proc");
    return true;
  }
  if (std::strcmp(proc, "button.pressed") == 0) {
    CallResult::ok(*ctx.reply);
    CallResult::appendKv(*ctx.reply, "lat", "50");
    CallResult::appendKv(*ctx.reply, "lon", "19");
    return true;
  }
  CallResult::ok(*ctx.reply);
  return true;
}

int main() {
  CommandRegistry reg;
  reg.registerCommand("ping", hPing);
  reg.registerCommand("call", hCall);
  // Registered features that are unavailable on this device (SPEC §18).
  reg.registerCommand("relay", hUnsupported);
  reg.registerCommand("battery", hUnsupported);
  reg.registerCommand("gps", hUnsupported);
  Dispatcher disp(reg);
  disp.setNodeName("tracker");
  disp.setGroupName("mych");

  fs::path dir = MCRPC_GOLDEN_DIR;
  if (!fs::exists(dir)) {
    std::printf("golden dir missing: %s\n", dir.c_str());
    return 1;
  }
  for (auto& ent : fs::directory_iterator(dir)) {
    if (ent.path().extension() != ".txt") continue;
    auto cases = loadFile(ent.path());
    for (auto& c : cases) runCase(ent.path().filename().c_str(), c, disp);
  }

  if (g_fail) {
    std::printf("GOLDEN: %d FAILED, %d passed\n", g_fail, g_pass);
    return 1;
  }
  std::printf("GOLDEN OK (%d assertions)\n", g_pass);
  return 0;
}
